File: python/src/semantic_retrieval/retrieval/csv_retriever.py
import pandas as pd
from result import Result
from semantic_retrieval.access_control.access_function import (
    AccessFunction,
    get_data_access_checked,
)
from semantic_retrieval.access_control.access_identity import (
    AuthenticatedIdentity,
)
from semantic_retrieval.common.types import CallbackEvent, Record
from semantic_retrieval.retrieval.retriever import BaseRetriever
from semantic_retrieval.utils.callbacks import CallbackManager, Traceable


class CSVRetrieverQuery(Record):
    # For now, assume a single primary key column
    primary_key_column: str


class CSVRetriever(BaseRetriever[pd.DataFrame, CSVRetrieverQuery], Traceable):
    """
    CSV retriever enforces RBAC at file granularity.
    If there is a need, this could be done at row-level.

    For simplicity, CSV retriever uses file path as the RBAC resource authentication ID.
    """

    def __init__(
        self,
        file_path: str,
        viewer_identity: AuthenticatedIdentity,
        user_access_function: AccessFunction,
        callback_manager: CallbackManager,
    ):
        super().__init__()
        self.file_path = file_path
        self.viewer_identity = viewer_identity
        self.user_access_function = user_access_function
        self.callback_manager = callback_manager

    async def retrieve_data(  # type: ignore
        self,
    ) -> Result[pd.DataFrame, str]:
        def _get_data(file_path: str) -> pd.DataFrame:
            return pd.read_csv(file_path).fillna(-1)

        out = await get_data_access_checked(
            self.file_path,
            self.user_access_function,
            _get_data,
            self.file_path,
            self.viewer_identity.viewer_auth_id,
        )

        await self.callback_manager.run_callbacks(
            CallbackEvent(
                name="csv_retriever_retrieved_data",
                data=dict(
                    file_path=self.file_path,
                    result=out,
                ),
            )
        )

        return out

File: typescript/src/retrieval/csvRetriever.ts
import { AccessPassport, ResourceAccessPolicy } from "../access-control";
import { JSONObject } from "../common/jsonTypes";
import { RetrieveDataEvent } from "../utils/callbacks";
import { BaseRetriever, BaseRetrieverQueryParams } from "./retriever";
import fs from "fs/promises";
import Papa from "papaparse";

export type CSVRetrieverQuery = {
  // For now, assume a single primary key column
  primaryKeyColumn: string;
};

export type CSVRetrieverConfig = {
  path: string;
  sourceAccessPolicies: ResourceAccessPolicy[];
};

export interface CSVRetrieverQueryParams
  extends BaseRetrieverQueryParams<CSVRetrieverQuery> {
  accessPassport: AccessPassport;
}

/**
 * Retrieve structured data from a CSV file as a JSON object for use in completion generation.
 */
export class CSVRetriever<R extends JSONObject> extends BaseRetriever {
  filePath: string;
  sourceAccessPolicies: ResourceAccessPolicy[];

  constructor(config: CSVRetrieverConfig) {
    super();
    this.filePath = config.path;
    this.sourceAccessPolicies = config.sourceAccessPolicies;
  }

  private async checkSourceAccess(
    params: CSVRetrieverQueryParams
  ): Promise<{ canAccess: boolean; policyFailed?: ResourceAccessPolicy }> {
    // For now, just check access at the source (file) level; row-level policies can be added
    // later if needed
    const policyChecks = await Promise.all(
      this.sourceAccessPolicies.map(async (policy) => {
        const identity = params.accessPassport.getIdentity(policy.resource);
        return {
          policy,
          check: identity && (await policy.testPolicyPermission(identity)),
        };
      })
    );

    // Default to hidden content. If no policies, or any policy fails, return null.
    if (policyChecks.length === 0) {
      return { canAccess: false };
    }

    for (const { policy, check } of policyChecks) {
      if (!check) {
        return { canAccess: false, policyFailed: policy };
      }

      if (Array.isArray(check)) {
        if (check.includes("read") || check.includes("*")) {
          continue;
        }
      }

      return { canAccess: false, policyFailed: policy };
    }

    return { canAccess: true };
  }

  /**
   * Get the data relevant to the given query and which the current identity can access.
   * If the current identity does not have access to the data, return null.
   * @param params The retriever query params to use for the query.
   * @returns A promise that resolves to the retrieved data, or null if no data accessible.
   */
  async retrieveData(params: CSVRetrieverQueryParams): Promise<R | null> {
    const accessCheck = await this.checkSourceAccess(params);
    if (!accessCheck.canAccess) {
      await this.callbackManager?.runCallbacks({
        name: "onRetrieverSourceAccessPolicyCheckFailed",
        params,
        policy: accessCheck.policyFailed ?? null,
      });
      return null;
    }

    const csvString = await (await fs.readFile(this.filePath)).toString();
    const rows = await Papa.parse<R>(csvString, {
      header: true,
      dynamicTyping: true,
    }).data;

    const retrievedData: JSONObject = {};
    const primaryKeyColumn = params.query.primaryKeyColumn;

    for (const row of rows) {
      const { [primaryKeyColumn]: key, ...restRow } = row;
      if (key == null) {
        if (restRow != null) {
          throw new Error(
            `Primary key column ${primaryKeyColumn} missing from row`
          );
        }
        continue; // ignore completely empty rows
      }

      const keyVal = typeof key === "string" ? key : JSON.stringify(key);
      retrievedData[keyVal] = restRow;
    }

    const event: RetrieveDataEvent = {
      name: "onRetrieveData",
      params,
      data: retrievedData,
    };

    await this.callbackManager?.runCallbacks(event);

    return retrievedData as R;
  }
}
